man/seaudit-report.8
.TH seaudit-report 8
.SH NAME
seaudit-report \- SELinux audit log reporting tool
.SH SYNOPSIS
.B seaudit-report
[OPTIONS] LOGFILES
.SH DESCRIPTION
.PP
.B seaudit-report
allows the user to generate custom audit log reports from the command line or by integration with the Logwatch tool.
.SH OPTIONS
.IP "-s, --stdin"
Read LOGFILES from standard input.
.IP "-m, --malformed"
Include malformed log messages.
.IP "-o FILE, --output=FILE"
Write output to FILE.
.IP "-c FILE, --config=FILE"
Use alternate config file.
.IP "--html"
Set output format to HTML.  Default is plain text.
.IP "--stylesheet=FILE"
HTML stylesheet for formatting HTML report.  This option is ignored if
--html is not given.  See /usr/local/share/setools/seaudit-report.css
(or wherever you installed SETools) for an example stylesheet.
.IP "-v, --version"
Print version information and exit.
.IP "-h, --help"
Print help information and exit.
.SH AUTHOR
This manual page was written by Jeremy A. Mowery <jmowery@tresys.com>.  
.SH COPYRIGHT
Copyright(C) 2004-2007 Tresys Technology, LLC
.SH SEE ALSO
seaudit(8)


NEWS
SETools 3.1:

SETools:

  * All tools that open a policy now support loadable policy modules.
    Command line tools expect the first module to be a base module
    followed optionally by any other modules.

libapol:

  * New class apol_policy_path_t to represent a base policy and any
    number of modules.  Use this whenever referring to the file or
    files constituting a policy.
  
libqpol:

  * Policy features such as attribute names or MLS can now be queried
    individally via qpol_policy_has_capability() rather than inferred
    by policy type and version.

libseaudit:

  * Rewrite of library to have proper namespaces.  libseaudit is now
    suitable for third-party users.

seaudit:

  * Rewrite to use new libseaudit.

  * Numerous tweaks to the interface to be more user friendly.

seaudit-report:

  * Rewrite to use new libseaudit.

sediffx:

  * Numerous tweaks to the interface to be more user friendly.

==================================================

SETools 3.0.1

SETools:

  * All code has been indented uniformly via the 'make indent' target.

SETools libraries:

  * All libraries now have a get_version() function.

libqpol:

  * Syntactic rule table is now off by default; it requires an
    explicit call to qpol_policy_build_syn_rule_table() to create it.

libapol:

  * apol_policy_t is now an opaque structure.  apol_permmap_t is no
    longer a public declaration.

  * avrule and terule queries now have full "syntactic" searching
    features.

apol & sesearch:

  * If loaded policy is source, the new syntactic search algorithm is used.

==================================================

SETools 3.0

SETools:

  * Moved entire project to autoconf build system.  This will detect
    dependencies correctly and make it easier to integrate with Linux
    distributions and their packaging systems.

SETools libraries:

  * Rewrite of back-end of SETools to use libsepol data structures.
    Most things should run notably faster.

  * All exported library functions have a standardized naming
    convention and are fully documented.  This will prove helpful for
    third-party developers integrating SETools into their own
    projects.

  * For policies version 15 or greater, domain transition analysis now
    takes into consideration setexec permission and/or type_transition
    rules.

apol:

  * Rework Tk interface to fit on 1024x768 displays.

  * Indirect matching of attributes now works with rule searches.

sediff:

  * New diffing algorithm yields significant speed improvements.

  * Can now diff neverallow and role_transition statements.

  * Streamlined results display - should be easier to read.

sechecker:

  * Updated module format and template for ease of extension.


=======================================================
May 1, 2006, Version 2.4

apol:
	File contexts tab now allows for MLS range searching if 
	the loaded database is from a MLS filesystem.
	Policy statistics dialog now shows MLS and ocontexts
	summaries.

libapol:
	Added support for loading base policies containing optionals.
	Added support for searching range transitions containing
	attributes.

libseaudit:
	Bugfix to support parsing FC5-style audit logs.

seaudit:
	Added date filters.

secmds:
	Added support to indexcon and searchcon for MLS filesytems.
	Added support to findcon and replcon for MLS filesystems.
	
sechecker:
	Added incomplete network access (inc_net_access) module.
	Added unreachable domains (unreachable_doms) module.
	Added impossible range transitions (imp_range_trans) module.

sesearch:
	Allow user to search range transitions by attributes and
	indirect matching.
	Added RBAC searching.


=======================================================
January 23, 2006, Version 2.3

apol:
	added new MLS components tab for sensitivities, 
	levels, and categories.
	changed users tab to support ranges and default 
	levels.
	added range transition tab for searching range
	transition rules.
	added new tab for network context components.
	added new tab for file system context components.
libapol:
	added binpol support for MLS, network contexts, 
	and file system contexts.
seinfo:
	added command line options for MLS components.
	added command line options for network contexts
	and file system contexts.
sesearch:
	added command line option for searching for rules
	by conditional boolean name.
seaudit:
	added new column in the log view for the 'comm' 
	field found in auditd log files.
	added filters for the 'comm' field and 'message'
	field.
manpages:
	added manpages for all tools.	
	
=======================================================
October 31, 2005, Version 2.2

libapol:
	replaced the original dta algorithm with a new one 
	to properly support complements in rules.  added
	new structures to support the separation of diff
	elements.  added support for parsing additional 
	policy components in source policies.
sediff:
	enhanced the GUI for display and separation of diff
	elements.  added the ability to rename types.
sechecker:
	added a new tool - a commandline modular and 
	extensible policy checker program
seuser:
	removed - deprecated
sepcut:
	removed - deprecated

=======================================================
October 12, 2005 Version 2.1.3

libapol:
	fixed a mls bug in the source parser.

=======================================================
August 24, 2005 Version 2.1.2

apol:
	created new permission maps for policy versions
	19 and 20.  also some minor changes to support
	version 20 binary format.
libseaudit:
	updated the parser to properly parse avc 
	messages from auditd logfiles
libapol:
	updated the binary policy parser to handle the
	new version 20 avtab format. The parser 
	preserves attributes in av rules by generating 
	fake attribute names.

=======================================================
May 17, 2005, Version 2.1.1

libseaudit:
	updated code to compile with gcc-4.0.0
	minor bug fixes

sediff:
	updated code to compile with gcc-4.0.0

seaudit:

seaudit/seaudit-report.c
/**
 *  @file seaudit-report.c
 *  Command line tool for processing SELinux audit logs and generating
 *  a concise report containing standard information as well as
 *  customized information using seaudit views.  Reports are rendered
 *  in either HTML or plain text.  Future support will provide
 *  rendering into XML.  The HTML report can be formatted by providing
 *  an alternate stylesheet file or by configuring the default
 *  stylesheet.  This tool also provides the option for including
 *  malformed strings within the report.
 *
 *  @author Jeremy A. Mowery jmowery@tresys.com
 *  @author Jason Tang jtang@tresys.com
 *
 *  Copyright (C) 2004-2006 Tresys Technology, LLC
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <config.h>

#include <seaudit/log.h>
#include <seaudit/parse.h>
#include <seaudit/report.h>

#include <apol/vector.h>

#include <libxml/xmlreader.h>

#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define COPYRIGHT_INFO "Copyright (C) 2004-2007 Tresys Technology, LLC"

static struct option const longopts[] = {
	{"html", no_argument, NULL, 'H'},
	{"malformed", no_argument, NULL, 'm'},
	{"output", required_argument, NULL, 'o'},
	{"stylesheet", required_argument, NULL, 'S'},
	{"stdin", no_argument, NULL, 's'},
	{"config", no_argument, NULL, 'c'},
	{"help", no_argument, NULL, 'h'},
	{"version", no_argument, NULL, 'v'},
	{NULL, 0, NULL, 0}
};

/**
 * Vector of seaudit_log_t, corresponding to each of the log files to
 * process.
 */
static apol_vector_t *logs = NULL;

/**
 * Error reporting log handler.
 */
static seaudit_log_t *first_log = NULL;

/**
 * Model that incorporates all of the logs within the logs vector.
 */
static seaudit_model_t *model = NULL;

/**
 * Report object for the above model.
 */
static seaudit_report_t *report = NULL;

static void seaudit_report_info_usage(const char *program_name, int brief)
{
	printf("%s (seaudit-report ver. %s)\n\n", COPYRIGHT_INFO, VERSION);
	printf("\nDescription: Generate a customized SELinux log report.\n");
	printf("Usage: %s [OPTIONS] LOGFILE ...\n", program_name);
	if (brief) {
		printf("\n   Try %s --help for more help.\n\n", program_name);
		return;
	}
	printf("  -s, --stdin              read log data from standard input\n");
	printf("  -m, --malformed          include malformed log messages\n");
	printf("  -oFILE, --output=FILE    output to FILE\n");
	printf("  -cFILE, --config=FILE    read configuration from FILE\n");
	printf("  --html                   set output format to HTML\n");
	printf("  --stylesheet=FILE        HTML style sheet for formatting HTML report\n");
	printf("                           (ignored if --html is not given)\n");
	printf("  -v, --version            print version information and exit\n");
	printf("  -h, --help               print this help text and exit\n");
	printf("\n");
	printf("Default style sheet is at %s.\n", APOL_INSTALL_DIR);
}

static void parse_command_line_args(int argc, char **argv)
{
	int optc, i;
	int do_malformed = 0, do_style = 0, read_stdin = 0;
	seaudit_report_format_e format = SEAUDIT_REPORT_FORMAT_TEXT;
	char *outfile = NULL, *configfile = NULL, *stylesheet = NULL;

	/* get option arguments */
	while ((optc = getopt_long(argc, argv, "o:c:t:msvh", longopts, NULL)) != -1) {
		switch (optc) {
		case 0:
			break;
		case 'o':	       /* output file name */
			outfile = optarg;
			break;
		case 'c':	       /* Alternate config file path */
			configfile = optarg;
			break;
		case 'S':	       /* HTML style sheet file path */
			stylesheet = optarg;
			do_style = 1;
			break;
		case 'm':	       /* include malformed messages */
			do_malformed = 1;
			break;
		case 's':	       /* read LOGFILES from standard input */
			read_stdin = 1;
			break;
		case 'H':	       /* Set the output to format to html */
			format = SEAUDIT_REPORT_FORMAT_HTML;
			do_style = 1;
			break;
		case 'v':
			/* display version */
			printf("\n%s (seaudit-report ver. %s)\n\n", COPYRIGHT_INFO, VERSION);
			exit(0);
		case 'h':
			/* display help */
			seaudit_report_info_usage(argv[0], 0);
			exit(0);
		default:
			/* display usage and handle error */
			seaudit_report_info_usage(argv[0], 1);
			exit(-1);
		}
	}

	/* Throw warning if a stylesheet was specified, but the --html
	 * option was not. */
	if (stylesheet != NULL && format != SEAUDIT_REPORT_FORMAT_HTML) {
		fprintf(stderr, "Warning: The --html option was not specified.\n");
		exit(-1);
	}

	if (!read_stdin && optind >= argc) {
		/* display usage and handle error */
		seaudit_report_info_usage(argv[0], 1);
		exit(-1);
	}

	if ((model = seaudit_model_create("seaudit-report", NULL)) == NULL) {
		exit(-1);
	}
	if ((first_log = seaudit_log_create(NULL, NULL)) == NULL || seaudit_model_append_log(model, first_log) < 0) {
		fprintf(stderr, "ERROR: %s\n", strerror(errno));
		exit(-1);
	}
	if ((logs = apol_vector_create()) == NULL || apol_vector_append(logs, first_log) < 0) {
		fprintf(stderr, "ERROR: %s\n", strerror(errno));
		exit(-1);
	}
	if (read_stdin) {
		/* Ensure that logfiles were not specified in addition
		 * to the standard-in option */
		if (optind < argc) {
			fprintf(stderr, "WARNING: %s\n", "Command line filename(s) will be ignored. Reading from stdin.");
		}
		if (seaudit_log_parse(first_log, stdin) < 0) {
			exit(-1);
		}
	} else {
		/* Parse given filenames */
		FILE *f;
		seaudit_log_t *l;
		if ((f = fopen(argv[optind], "r")) == NULL) {
			fprintf(stderr, "ERROR: %s\n", strerror(errno));
			exit(-1);
		}
		if (seaudit_log_parse(first_log, f) < 0) {
			exit(-1);
		}
		fclose(f);
		for (i = optind + 1; i < argc; i++) {
			if ((l = seaudit_log_create(NULL, NULL)) == NULL || seaudit_model_append_log(model, l) < 0) {
				exit(-1);
			}
			if (apol_vector_append(logs, l) < 0) {
				fprintf(stderr, "ERROR: %s\n", strerror(errno));
				exit(-1);
			}
			if ((f = fopen(argv[i], "r")) == NULL) {
				fprintf(stderr, "ERROR: %s\n", strerror(errno));
				exit(-1);
			}
			if (seaudit_log_parse(l, f) < 0) {
				exit(-1);
			}
			fclose(f);
		}
	}

	if ((report = seaudit_report_create(model, outfile)) == NULL ||
	    seaudit_report_set_format(first_log, report, format) < 0 ||
	    seaudit_report_set_configuration(first_log, report, configfile) < 0 ||
	    seaudit_report_set_stylesheet(first_log, report, stylesheet, do_style) < 0 ||
	    seaudit_report_set_malformed(first_log, report, do_malformed) < 0) {
		exit(-1);
	}
}

int main(int argc, char **argv)
{
	size_t i;
	parse_command_line_args(argc, argv);
	if (seaudit_report_write(first_log, report) < 0) {
		return -1;
	}
	seaudit_report_destroy(&report);
	seaudit_model_destroy(&model);
	for (i = 0; i < apol_vector_get_size(logs); i++) {
		seaudit_log_t *l = apol_vector_get_element(logs, i);
		seaudit_log_destroy(&l);
	}
	apol_vector_destroy(&logs, NULL);
	return 0;
}
